csharp/src/Google.Protobuf/SegmentedBufferHelper.cs

```diff
@@ -269,5 +269,22 @@ private static void CheckCurrentBufferIsEmpty(ref ParserInternalState state)
                 throw new InvalidOperationException("RefillBuffer() called when buffer wasn't empty.");
             }
         }
+
+        private static long? ResolveTotalLength(CodedInputStream stream)
+        {
+            if (stream.InternalInputStream == null)
+            {
+                return stream.InternalBuffer.Length;
+            }
+
+            // By convention, if CanSeek is true, then the Length should be available.
+            if (stream.InternalInputStream.CanSeek)
+            {
+                return stream.InternalInputStream.Length;
```

Comment:
For non-BCL Stream, this could still throw which would be a small breaking change. Is that acceptable?